File: src/main/dev/danzou/advent22/scala/Day3.scala
package dev.danzou.advent22.scala

import dev.danzou.advent.utils.AdventTestRunner

class Day3 extends AdventTestRunner {
  private def priorityOf(char: Char): Int = {
    if (char.isUpper) char - 'A' + 27
    else char - 'a' + 1
  }

  override def part1(input: String): Number =
    input.split("\n")
      .map(str => str.toList)
      .map(list =>
        list.slice(0, list.size / 2).intersect(list.slice(list.size / 2, list.size)))
      .map(list => priorityOf(list.head))
      .sum

  override def part2(input: String): Number = ???
}

File: src/main/dev/danzou/advent22/kotlin/Day3.kt
package dev.danzou.advent22.kotlin

import dev.danzou.advent22.AdventTestRunner22

internal class Day3 : AdventTestRunner22() {
    fun priorityOf(char: Char): Int =
        if (char.isUpperCase()) char - 'A' + 27
        else char - 'a' + 1

    override fun part1(input: String): Number {
        return input.split("\n")
            .map { it.toList() }
            .map { it.chunked(it.size / 2).reduce(Iterable<Char>::intersect) }
            .map { priorityOf(it.first()) }
            .sum()
    }

    override fun part2(input: String): Number {
        return input.split("\n")
            .map { it.toList() }
            .chunked(3)
            .map { it.reduce(Iterable<Char>::intersect) }
            .map { priorityOf(it.first()) }
            .sum()
    }
}